File: src/RankOne.SEO.Tool/Web/UI/App_Plugins/RankOne/services/score.service.js
angular.module('umbraco')
    .service('scoreService',
        function () {

            function getScoreForResultInternal(result) {
                var score = 100;
                if (result) {

                    if (result.errorCount > 0) {
                        score = 0;
                    }
                    score -= result.warningCount * 50;
                    score -= result.hintCount * 25;

                    if (score < 0) {
                        score = 0;
                    }
                }
                return score;
            }


            this.getScoreForResult = function (result) {
                return getScoreForResultInternal(result);
            }
            this.getOverallScore = function (results) {
                var combinedScore = 0;
                angular.forEach(results, function (result) {
                    combinedScore += getScoreForResultInternal(result);
                });
                return Math.round(combinedScore / results.length);
            }

            this.getTotalErrorCount = function (results) {
                var combinedCount = 0;
                angular.forEach(results, function (result) {
                    combinedCount += result.errorCount;
                });
                return combinedCount;
            }

            this.getTotalWarningCount = function (results) {
                var combinedCount = 0;
                angular.forEach(results, function (result) {
                    combinedCount += result.warningCount;
                });
                return combinedCount;
            }

            this.getTotalHintCount = function (results) {
                var combinedCount = 0;
                angular.forEach(results, function (result) {
                    combinedCount += result.hintCount;
                });
                return combinedCount;
            }

            this.getTotalSuccessCount = function (results) {
                var combinedCount = 0;
                angular.forEach(results, function (result) {
                    combinedCount += result.successCount;
                });
                return combinedCount;
            }
        });

File: src/RankOne.SEO.Tool/Services/ScoreService.cs
using RankOne.Interfaces;
using RankOne.Models;
using System;
using System.Collections.Generic;
using System.Linq;

namespace RankOne.Services
{
    public class ScoreService : IScoreService
    {
        public PageScore GetScore(PageAnalysis pageAnalysis)
        {
            if (pageAnalysis == null) throw new ArgumentNullException(nameof(pageAnalysis));

            var analyzers = pageAnalysis.SummaryResults.SelectMany(x => x.Analysis.Results);

            return CalculatePageScore(analyzers);
        }

        private PageScore CalculatePageScore(IEnumerable<AnalyzeResult> analyzers)
        {
            var pageScore = new PageScore
            {
                ErrorCount = analyzers.Sum(x => x.ErrorCount),
                WarningCount = analyzers.Sum(x => x.WarningCount),
                HintCount = analyzers.Sum(x => x.HintCount),
                SuccessCount = analyzers.Sum(x => x.SuccessCount)
            };
            var totalScore = analyzers.Sum(CalculateResultScore);
            var totalWeight = analyzers.Sum(x => x.Weight);

            pageScore.OverallScore = CalculateOverallScore(totalScore, totalWeight);

            return pageScore;
        }

        private int CalculateOverallScore(double totalScore, double totalWeight)
        {
            return (int)Math.Round(totalScore / (totalWeight / 100.0));
        }

        /// <summary>
        /// Gets the result score per analyzer.
        /// </summary>
        /// <param name="result">The result.</param>
        /// <returns>Score for the analyzer</returns>
        private int CalculateResultScore(AnalyzeResult result)
        {
            var score = 100;

            // If there are any errors, the score is 0
            if (result.ErrorCount > 0) return 0;

            // Each warning costs 50%
            score -= result.WarningCount * 50;

            // Each hint costs 25 %
            score -= result.HintCount * 25;
            if (score < 0) return 0;
            return score;
        }
    }
}